### PR TITLE
[merged] Support a configuration file

### PR DESCRIPTION
**Pull request by mbarnes**
The configuration file format is JSON (because I'm lazy) and I picked a default path of `/etc/commissaire/commissaire.conf`, which `--config-file=PATH` will override.

Note the changes to the `Dockerfile` and the docs for running in a container.  I wrote it so it expects a config file mounted at the default path, but even with `:Z` on the volume option I still had to fight with SELinux.  Now I'm wondering if a better practice would be to build a base image from our `Dockerfile` and then add a site-specific config file to a derivative image.

Also, this line in the documented sample config looks pretty clunky now:
```
"authentication-plugin-kwargs": "filepath=conf/users.json"
```
Would be nicer to allow for that user data inline.  (tech debt?)

### CODE REVIEW

**Comment by ashcrow** [comment]
Reviewing now.

**Comment by ashcrow** [comment]
> The configuration file format is JSON (because I'm lazy) 

Most everything else right now is JSON in commissaire so I think this is actually a good idea.

> Would be nicer to allow for that user data inline. (tech debt?)

I don't follow what you mean by 'inline'. Can you explain?

File: src/commissaire/script.py [hunk 1]
```diff
@@ -159,9 +160,25 @@ def main():  # pragma: no cover
         required=False, default={},
         help='Authentication Plugin configuration (key=value)')
 
-    args = parser.parse_args()
+    # XXX Temporary
```
Comment (ashcrow) [inline src/commissaire/script.py]: Just wondering, why is this marked temporary?

File: src/commissaire/script.py [hunk 2]
```diff
@@ -124,44 +166,66 @@ def parse_args(parser):
...
         default='commissaire.authentication.httpauthbyetcd',
         help=('Authentication Plugin module. '
               'EX: commissaire.authentication.httpauth'))
     parser.add_argument(
-        '--authentication-plugin-kwargs', type=str,
-        required=False, default={},
+        '--authentication-plugin-kwargs', type=str, default={},
         help='Authentication Plugin configuration (key=value)')
 
-    # XXX Temporary
```
Comment (ashcrow) [inline src/commissaire/script.py]: Well, never mind :smile:

File: src/commissaire/script.py [hunk 3]
```diff
@@ -111,6 +111,48 @@ def parse_uri(uri, name):
...
+
+    :param path: Full path to the config file, or None
+    :type path: str or None
+    :returns: A namespace object, possibly empty
+    :rtype: argparse.Namespace
+    """
+    json_object = {}
+    using_default = False
+
+    if path is None:
```
Comment (ashcrow) [inline src/commissaire/script.py]: nit: It would be nice to log that we are falling back to the default config path. Not required, but nice to have.

**Comment by ashcrow** [comment]
I'm running BDD tests right now. I noticed one failure. I'll report back once they finish.

**Comment by ashcrow** [comment]
```
Failing scenarios:
  features/cluster/hosts.feature:179  Deleting a host also removes it from its cluster (1)
  features/cluster/hosts.feature:191  Deleting a host also removes it from its cluster (2)
  features/cluster/hosts.feature:203  Deleting a host also removes it from its cluster (3)
```

I'm running on master to see if this introduced the break or it was already there.

**Comment by ashcrow** [comment]
This change didn't introduce the failures ... so it's fine from that POV as well.

@mbarnes do you want to add in the dict/str to plugins or open a card for a future enhancement?

**Comment by mbarnes** [comment]
> @mbarnes do you want to add in the dict/str to plugins or open a card for a future enhancement?

I don't want it to block this PR in case I hit complications.  I'll open a separate card, but still have a go at hammering it out quickly.

Any comment about the `Dockerfile` changes I made?  Didn't know if I should be concerned about backward-compatibility.  Will the invocation changes break things for @cooktheryan?

I'll get that log message you wanted shortly here.

**Comment by ashcrow** [comment]
@cooktheryan thoughts? ^

**Comment by mbarnes** [comment]
:arrow_up:
That was a little trickier than you would think because the logging configuration is not yet loaded at that point and info-level log message are eaten by default.  After trying a few things I said "screw it, just print."

**Comment by mbarnes** [comment]
:arrow_up: Tacked on a trivial fix for the failing test case, since you mentioned it here.

**Comment by ashcrow** [comment]
LGTM

@cgwalters-bot try

**Comment by ashcrow** [comment]
@cgwalters-bot r+

**Pull request closed without merging**